### ISSUE TITLE
GitHubAuth: 'NoneType' object has no attribute '__getitem__'

### ISSUE THREAD

**Issue by ibuclaw** [buildbot/buildbot#4077]
Config:
```
        c['www']['auth'] = util.GitHubAuth(os.environ['BUILDMASTER_CLIENT_ID'],
                                           os.environ['BUILDMASTER_CLIENT_SECRET'],
                                           apiVersion=4, getTeamsMembership=True)
```

Traceback when attempting to login using oauth.
```
buildbot_1      | 2018-05-05 10:25:48+0000 [-] /usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: urllib3.exceptions.InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
buildbot_1      | 2018-05-05 10:25:49+0000 [-] /usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: urllib3.exceptions.InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
buildbot_1      | 2018-05-05 10:25:49+0000 [-] While rendering resource:
buildbot_1      | 	Traceback (most recent call last):
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 434, in errback
buildbot_1      | 	    self._startRunCallbacks(fail)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
buildbot_1      | 	    self._runCallbacks()
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
buildbot_1      | 	    current.result = callback(current.result, *args, **kw)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1184, in gotResult
buildbot_1      | 	    _inlineCallbacks(r, g, deferred)
buildbot_1      | 	--- <exception caught here> ---
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
buildbot_1      | 	    result = result.throwExceptionIntoGenerator(g)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
buildbot_1      | 	    return g.throw(self.type, self.value, self.tb)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 63, in renderLogin
buildbot_1      | 	    details = yield self.auth.verifyCode(code)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 246, in inContext
buildbot_1      | 	    result = inContext.theWork()
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 262, in <lambda>
buildbot_1      | 	    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
buildbot_1      | 	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
buildbot_1      | 	    return func(*args,**kw)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 173, in thd
buildbot_1      | 	    return self.getUserInfoFromOAuthClient(session)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 276, in getUserInfoFromOAuthClient
buildbot_1      | 	    return self.getUserInfoFromOAuthClient_v4(c)
buildbot_1      | 	  File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 340, in getUserInfoFromOAuthClient_v4
buildbot_1      | 	    teams.add('%s/%s' % (orgs_name_slug_mapping[org], node['node']['name']))
buildbot_1      | 	exceptions.TypeError: 'NoneType' object has no attribute '__getitem__'
buildbot_1      | 	
```

The OAuth app in github has:
-  Read org and team membership
- Access user email addresses (read-only)

**Comment by ibuclaw** [comment]
One of the organisations I'm a part of restricts access to third party apps, and in the debug json output the following content is returned:
```
{u'data': {u'org1': {u'teams': {u'edges': [{u'node': None},
                                                        {u'node': None},
                                                        {u'node': None},
                                                        {u'node': None},
                                                        {u'node': None}]}},
           u'org2': {u'teams': {u'edges': [{u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None},
                                            {u'node': None}]}},
           u'org3': {u'teams': {u'edges': [{u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None},
                                                {u'node': None}]}}},
 u'errors': [{u'locations': [{u'column': 9, u'line': 6}],
              u'message': u'Although you appear to have the correct authorization credentials,\nthe `org2` organization has enabled OAuth App access restrictions, meaning that data\naccess to third-parties is limited. For more information on these restrictions, including\nhow to whitelist this app, visit\nhttps://help.github.com/articles/restricting-access-to-your-organization-s-data/\n',
              u'path': [u'org2', u'teams', u'edges', 0, u'node'],
              u'type': u'FORBIDDEN'}]}
```
(There are lots of errors, but removed for brevity).

**Comment by tardyp** [comment]
Hi,

thanks for investigation!

would you like to send a PR doing the check for None in oauth2.py line 340?

**Comment by ibuclaw** [comment]
@tardyp - Yes, will do.

**Comment by ibuclaw** [comment]
FYI, github has fixed their bug in GraphQL.  Organizations that don't restrict third party access will now correctly return team information.

Might need to update the fix in #4096 though.

The returned data I get now is:
```
{   u'data': {   u'org1': {   u'teams': {   u'edges': [   {   u'node': {   u'name': u'Team One',
                                                                                        u'slug': u'team1'}},
                                                                       {   u'node': {   u'name': u'Team Two',
                                                                                        u'slug': u'team2'}}]}},
                 u'org2': None,
                 u'org3': {   u'teams': {   u'edges': [   {   u'node': {   u'name': u'Team One',
                                                                                u'slug': u'team1'}},
                                                               {   u'node': {   u'name': u'Team Two',
                                                                                u'slug': u'team2'}}]}}},
    u'errors': [   {   u'locations': [{   u'column': 5, u'line': 4}],
                       u'message': u'Although you appear to have the correct authorization credentials,\nthe `org2` organization has enabled OAuth App access restrictions, meaning that data\naccess to third-parties is limited. For more information on these restrictions, including\nhow to whitelist this app, visit\nhttps://help.github.com/articles/restricting-access-to-your-organization-s-data/\n',
                       u'path': [u'org2', u'teams'],
                       u'type': u'FORBIDDEN'}]}
```

**Comment by ibuclaw** [comment]
Infact, yes.  The beginning of the backtrace has now moved to the for loop.
```
| 2018-05-25 18:39:08+0000 [-] /usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: urllib3.exceptions.InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
| 2018-05-25 18:39:08+0000 [-] /usr/lib/python2.7/site-packages/urllib3/connectionpool.py:858: urllib3.exceptions.InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
| 2018-05-25 18:39:08+0000 [-] While rendering resource:
|         Traceback (most recent call last):
|           File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 434, in errback
|             self._startRunCallbacks(fail)
|           File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
|             self._runCallbacks()
|           File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
|             current.result = callback(current.result, *args, **kw)
|           File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1184, in gotResult
|             _inlineCallbacks(r, g, deferred)
|         --- <exception caught here> ---
|           File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
|             result = result.throwExceptionIntoGenerator(g)
|           File "/usr/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
|             return g.throw(self.type, self.value, self.tb)
|           File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 63, in renderLogin
|             details = yield self.auth.verifyCode(code)
|           File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 246, in inContext
|             result = inContext.theWork()
|           File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 262, in <lambda>
|             inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
|           File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
|             return self.currentContext().callWithContext(ctx, func, *args, **kw)
|           File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
|             return func(*args,**kw)
|           File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 173, in thd
|             return self.getUserInfoFromOAuthClient(session)
|           File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 276, in getUserInfoFromOAuthClient
|             return self.getUserInfoFromOAuthClient_v4(c)
|           File "/usr/lib/python2.7/site-packages/buildbot/www/oauth2.py", line 334, in getUserInfoFromOAuthClient_v4
|             for node in team_data['teams']['edges']:
|         exceptions.TypeError: 'NoneType' object has no attribute '__getitem__'
```
